experimental/hip/pipeline_layout.c

```diff
@@ -0,0 +1,258 @@
+        push_constant_count, IREE_HAL_HIP_MAX_PUSH_CONSTANT_COUNT);
+  }
+
+  // Currently the pipeline layout doesn't do anything.
+  // TODO: Handle creating the argument layout at that time hadling both push
+  // constant and buffers.
+  iree_hal_hip_pipeline_layout_t* pipeline_layout = NULL;
+  iree_host_size_t total_size =
+      sizeof(*pipeline_layout) +
+      set_layout_count * sizeof(*pipeline_layout->set_layouts);
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_allocator_malloc(host_allocator, total_size,
+                                (void**)&pipeline_layout));
+
+  iree_hal_resource_initialize(&iree_hal_hip_pipeline_layout_vtable,
+                               &pipeline_layout->resource);
+  pipeline_layout->host_allocator = host_allocator;
+  pipeline_layout->set_layout_count = set_layout_count;
+  iree_host_size_t base_index = 0;
+  for (iree_host_size_t i = 0; i < set_layout_count; ++i) {
+    pipeline_layout->set_layouts[i].set_layout = set_layouts[i];
+    // Copy and retain all descriptor sets so we don't lose them.
+    iree_hal_descriptor_set_layout_retain(set_layouts[i]);
+    pipeline_layout->set_layouts[i].base_index = base_index;
+    base_index +=
+        iree_hal_hip_descriptor_set_layout_binding_count(set_layouts[i]);
+  }
+  pipeline_layout->push_constant_base_index = base_index;
+  pipeline_layout->push_constant_count = push_constant_count;
+  *out_pipeline_layout = (iree_hal_pipeline_layout_t*)pipeline_layout;
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static void iree_hal_hip_pipeline_layout_destroy(
+    iree_hal_pipeline_layout_t* base_pipeline_layout) {
+  iree_hal_hip_pipeline_layout_t* pipeline_layout =
+      iree_hal_hip_pipeline_layout_cast(base_pipeline_layout);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_allocator_t host_allocator = pipeline_layout->host_allocator;
+
+  for (iree_host_size_t i = 0; i < pipeline_layout->set_layout_count; ++i) {
+    iree_hal_descriptor_set_layout_release(
+        pipeline_layout->set_layouts[i].set_layout);
+  }
+  iree_allocator_free(host_allocator, pipeline_layout);
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+iree_host_size_t iree_hal_hip_pipeline_layout_descriptor_set_count(
+    const iree_hal_pipeline_layout_t* base_pipeline_layout) {
+  const iree_hal_hip_pipeline_layout_t* pipeline_layout =
+      iree_hal_hip_pipeline_layout_const_cast(base_pipeline_layout);
+  return pipeline_layout->set_layout_count;
+}
+
+const iree_hal_descriptor_set_layout_t*
+iree_hal_hip_pipeline_layout_descriptor_set_layout(
+    const iree_hal_pipeline_layout_t* base_pipeline_layout, uint32_t set) {
+  const iree_hal_hip_pipeline_layout_t* pipeline_layout =
+      iree_hal_hip_pipeline_layout_const_cast(base_pipeline_layout);
+  if (set < pipeline_layout->set_layout_count) {
+    return pipeline_layout->set_layouts[set].set_layout;
+  }
+  return NULL;
+}
+
+iree_host_size_t iree_hal_hip_pipeline_layout_base_binding_index(
```

Comment:
instead of one function call per query it's better to make one function call that returns a struct with the information you need at the site you are requesting it (dispatch) - so returning a `iree_hal_hip_dispatch_layout_t` with your binding/push constant counts/indices/etc. optimizing for dispatch time is important and since the only part of the HAL that uses pipeline layouts is dispatch we should be building what's stored here and how it's retrieved to optimize for that and not put any additional API barriers in the way.

the goal is that when you go to dispatch you do literally nothing but fetch a struct and read some fields, maybe with one loop over data you couldn't have precached. pipeline layouts are intended to make dispatch faster as their primary goal by amortizing the cost of things like building bitmaps for packing bindings and such.

this means you really want to be writing pipeline layout code at the same time you are writing your dispatch code and not the other way around - if you don't yet know how you're going to dispatch you'd be better deleting all these accessor methods for now until you do.

Reply:
okay got it, I'm going to remove these functions for now and will add accordingly when doing the dispatch

Reply:
@antiagainst From what I see, what Ben said applies to cuda2 backend as well. Do we want to modify it?

Reply:
Yup, please feel free to fix whatever that applies to cuda2 side too!

Reply:
I have removed all accessor methods and constructed a struct `iree_hal_hip_dispatch_layout_t` with the parameters except these two. I refrained from adding these two methods because we'd have to redo a memory allocation to store the set layout parameters essentially duplicating the information in `iree_hal_pipeline_layout_t`. 

I was thinking if we expose `iree_hal_hip_pipeline_layout_t` itself in command_buffer then we have access to all these parameters after a single allocation which would be efficient. Is there any reason to not expose `iree_hal_hip_pipeline_layout_t` in command_buffer? I followed the new cuda implementation and I was wondering if there's a reason for that.